backends/p4tools/common/options.cpp

```diff
@@ -0,0 +1,226 @@
+
+    return boost::make_optional(::errorCount() == 0, compilerContext);
+}
+
+std::vector<const char*>* AbstractP4cToolOptions::process(int argc, char* const argv[]) {
+    return Util::Options::process(argc, argv);
+}
+
+/// Specifies a command-line option to inherit from the compiler, and any special handling for the
+/// option.
+struct InheritedCompilerOptionSpec {
+    /// The name of the command-line option. For example, "--target".
+    const char* option;
+
+    /// A descriptive name for the parameter to the option, or nullptr if the option has no
+    /// parameters.
+    const char* argName;
+
+    /// A description of the option.
+    const char* description;
+
+    /// An optional handler for the option. If provided, this is executed before the option is
+    /// forwarded to the compiler. Any argument to the option is provided to the handler. The
+    /// handler should return true on successful processing, and false otherwise.
+    boost::optional<std::function<bool(const char*)>> handler;
+};
+
+AbstractP4cToolOptions::AbstractP4cToolOptions(cstring message) : Options(message) {
+    // Register some common options.
+    registerOption(
+        "--help", nullptr,
+        [this](const char*) {
+            usage();
+            exit(0);
+            return false;
+        },
+        "Shows this help message and exits");
+
+    registerOption(
+        "--version", nullptr,
+        [this](const char*) {
+            printVersion(binaryName);
+            exit(0);
+            return false;
+        },
+        "Prints version information and exits");
+
+    registerOption(
+        "--min-packet-size", "bytes",
+        [this](const char* arg) {
+            int minLen_bytes = std::atoi(arg);
+            if (minLen_bytes <= 0) {
+                ::error("Invalid minimum packet length: %1%", arg);
+                return false;
+            }
+            minPacketSize_bytes = minLen_bytes;
+            return true;
+        },
+        "Sets the minimum allowed packet size, in bytes. Any packet shorter than this is "
+        "considered to be invalid, and will be dropped if the program would otherwise send the "
+        "packet on the network.");
+
+    registerOption(
+        "--mtu", "bytes",
+        [this](const char* arg) {
+            int maxLen_bytes = std::atoi(arg);
+            if (maxLen_bytes <= 0) {
+                ::error("Invalid network MTU: %1%", arg);
+                return false;
+            }
+            networkMtu_bytes = maxLen_bytes;
+            return true;
+        },
+        "Sets the network's MTU, in bytes");
+
+    registerOption(
+        "--seed", "seed",
+        [this](const char* arg) {
+            seed = std::stoul(arg);
+            return true;
+        },
+        "Provides a randomization seed");
+
+    // Inherit some compiler options, setting them up to be forwarded to the compiler.
+    std::vector<InheritedCompilerOptionSpec> inheritedCompilerOptions = {
```

Comment:
can't this be done by inheriting another Options class?

Reply:
I am actually not sure why this options class was written this way. Let me see.

Reply:
I tried simplifying this but unfortunately this requires a major rewrite. These options are written this way to make it possible to have every target extension define its own options without needing to produce a different binary. 
`p4check testgen`, `p4check mutate`,  `p4check verify` can all have their own options.